### PR TITLE
TSV-131: Throttle users to mandatory/optional 2SV

### PR DESCRIPTION
**Pull request by paoloboni**

### CODE REVIEW

File: app/services/TwoStepVerificationThrottle.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+package services
+
+import org.joda.time.DateTime
+import play.api.Play.{current, _}
+
+trait TwoStepVerificationThrottle {
+  def hourlyLimit: HourlyLimit
+
+  def registrationMandatory(discriminator: String) = {
+    val userValue = Math.abs(discriminator.hashCode % 100)
```
Comment (tamaslang) [inline app/services/TwoStepVerificationThrottle.scala]: I am wondering how likely the last 2 digits are random and providing a balance spread based on a scale 0-99 considering the impementation is: http://docs.oracle.com/javase/6/docs/api/java/lang/String.html#hashCode()

Comment (paoloboni) [inline app/services/TwoStepVerificationThrottle.scala]: Distribution should be uniform - tested with brute force

File: app/services/TwoStepVerificationThrottle.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    userValue <= threshold
+  }
+}
+
+object TwoStepVerificationThrottle extends TwoStepVerificationThrottle {
+  override lazy val hourlyLimit = HourlyLimit
+}
+
+
+trait HourlyLimit {
```
Comment (dorinp) [inline app/services/TwoStepVerificationThrottle.scala]: naming: HourlyLimit suggests that there is a limit applied per hour - which is not the case. I suggest calling it TimeBasedLimit or similar

File: app/services/TwoStepVerificationThrottle.scala (hunk 1 above)
Comment (tamaslang) [inline app/services/TwoStepVerificationThrottle.scala]: 👍

File: app/services/TwoStepVerification.scala [hunk 3]
```diff
@@ -55,14 +59,24 @@ trait TwoStepVerification {
           }
         }
         shouldRedirectTo2SV.map {
-          case true => Some(wrapLocationWith2SV(continue))
+          case true =>
+            twoStepVerificationThrottle.registrationMandatory(authContext.user.oid) match {
+              case true =>
+                val continueToAccountUrl = URLEncoder.encode(s"${ExternalUrls.taxAccountRouterHost}/account", "UTF-8")
```
Comment (tamaslang) [inline app/services/TwoStepVerification.scala]: Is there a business value with calculating it every time?

File: app/services/TwoStepVerificationThrottle.scala (hunk 1 above)
Comment (tamaslang) [inline app/services/TwoStepVerificationThrottle.scala]: 👍

File: app/services/TwoStepVerification.scala (hunk 3 above)
Comment (paoloboni) [inline app/services/TwoStepVerification.scala]: No, actually urls can be moved out of this and be constants

File: app/model/AuditContext.scala [hunk 4]
```diff
@@ -87,7 +87,18 @@ trait TAuditContext {
 
   private val routingReasons: mutableMap[String, String] = defaultRoutingReasons
   private val throttlingDetails: mutableMap[String, String] = mutableMap.empty
-  var sentTo2SVRegister = false
+  private var sentToOptional2SVRegister = false
```
Comment (anshulbajpai) [inline app/model/AuditContext.scala]: Convert two states to one state

File: app/services/TwoStepVerification.scala [hunk 5]
```diff
@@ -44,6 +48,9 @@ trait TwoStepVerification {
     BusinessTaxAccount -> "business-tax-account"
   )
 
+  val continueToAccountUrl = URLEncoder.encode(s"${ExternalUrls.taxAccountRouterHost}/account", "UTF-8")
```
Comment (anshulbajpai) [inline app/services/TwoStepVerification.scala]: Can be private?

File: app/services/TwoStepVerification.scala [hunk 6]
```diff
@@ -55,14 +62,22 @@ trait TwoStepVerification {
...
+            }
           case _ => None
         }
-      }.andThen { case Success(Some(_)) => auditContext.sentTo2SVRegister = true }
+      }
     } else Future.successful(None)
   }
 
-  private def wrapLocationWith2SV(continue: Location) = Locations.twoStepVerification(Map("continue" -> continue.fullUrl, "failure" -> continue.fullUrl) ++
+  private def wrapLocationWith2SV(continue: Location, failure: String) = Locations.twoStepVerification(Map("continue" -> continue.fullUrl, "failure" -> failure) ++
```
Comment (anshulbajpai) [inline app/services/TwoStepVerification.scala]: I understand this is a private method but can both the parameters be of the same type?

File: app/services/TwoStepVerificationThrottle.scala [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+
+trait TimeBasedLimit {
+  def dateTimeProvider: () => DateTime
+
+  def getCurrentPercentageLimit = {
+    val currentHourOfDay = dateTimeProvider().getHourOfDay
+    val hourlyLimit = configuration.getInt(s"two-step-verification.throttle.$currentHourOfDay")
+    val defaultLimit = configuration.getInt("two-step-verification.throttle.default")
```
Comment (anshulbajpai) [inline app/services/TwoStepVerificationThrottle.scala]: Can we precalculate default limit inside trait definition?

File: app/services/TwoStepVerificationThrottle.scala [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    hourlyLimit.orElse(defaultLimit).getOrElse(0)
```
Comment (anshulbajpai) [inline app/services/TwoStepVerificationThrottle.scala]: Can defaultLimit value itself be smart to default to 0 if configuration does not exist?

File: test/services/TimeBasedLimitSpec.scala [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    "return the hourly limit in the configuration for the current hour if defined" in new Setup {
+      val expectedLimit = 10
+      running(FakeApplication(additionalConfiguration = Map("two-step-verification.throttle.12" -> expectedLimit))) {
+        timeBasedLimit.getCurrentPercentageLimit shouldBe expectedLimit
+      }
+    }
+    "return the 'default' limit in configuration if hourly limit is not defined for current hour" in new Setup {
+      val expectedLimit = 15
+      running(FakeApplication(additionalConfiguration = Map(
+        "two-step-verification.throttle.12" -> null,
```
Comment (anshulbajpai) [inline test/services/TimeBasedLimitSpec.scala]: Can we have a number here instead of null?

**Pull request merged**